api/src/main/java/org/apache/iceberg/expressions/ResidualEvaluator.java

```diff
@@ -214,6 +214,11 @@ public Expression or(Expression leftResult, Expression rightResult) {
       return ((String) ref.eval(struct)).startsWith((String) lit.value()) ? alwaysTrue() : alwaysFalse();
     }
 
+    @Override
+    public <T> Expression notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      return ((String) ref.eval(struct)).startsWith((String) lit.value()) ? alwaysFalse() : alwaysTrue();
+    }
```

Comment:
This is another case where I chose to go with the existing style of the surrounding code (e.g. the `startsWith` definition just above this), but I'm not so sure that casting to java String type is appropriate vs using the built in comparators.

Comparing the values as ByteBuffers, given the differences in Java's UTF-16 default and the included Literals comparator which handles UTF-8 code point differences for characters that are encoded as 3 vs 4 bytes, seems safer to me but I chose to be safe and emulate the surrounding code.

Reply:
If we can't use String comparison, then the right way to compare is to convert to `CharSequence`. `ByteBuffer` gets converted to `CharBuffer` using UTF-8 and that implements `CharSequence`, while `String` already implements `CharSequence`.

We know that the literal's value is a `String` because that is the only literal value that is accepted by the API, so that cast is fine here and in other places. I'm not sure that the result of evaluating the reference is a String. It seems like that should be a `CharSequence` instead. But let's fix that problem later.

Reply:
Ok. I agree on fixing it later (if necessary). And the `CharSequence` recommendation seems much more sane, as that's all the API allows for and then we reduce allocations.

Thanks for the info, Ryan.

Reply:
I'm going to resolve this conversation, as we've decided to fix this later, if need be. This is to make the PR more digestible for others. Feel free to open it back up if you'd like.